pydeseq2/ds.py

```diff
@@ -681,31 +585,61 @@ def _build_contrast_vector(self) -> None:
-            self.contrast_idx = self.LFC.columns.get_loc(contrast_level)
-            self.contrast_vector[self.contrast_idx] = 1
-        elif f"{factor}_{ref}_vs_{alternative}" in self.design_matrix.columns:
-            # Reference and alternative are inverted
-            self.contrast_idx = self.LFC.columns.get_loc(
-                f"{factor}_{ref}_vs_{alternative}"
-            )
-            self.contrast_vector[self.contrast_idx] = -1
-        else:
-            # Need to change reference
-            # Get any column corresponding to the desired factor and extract old ref
-            old_ref = next(
-                col for col in self.LFC.columns if col.startswith(factor)
-            ).split("_vs_")[-1]
-            new_alternative_idx = self.LFC.columns.get_loc(
-                f"{factor}_{alternative}_vs_{old_ref}"
+        self.contrast_vector = self._contrast(
+            column=factor, baseline=ref, group_to_compare=alternative
+        )
+
+    # Everything below is copied from pertpy. TODO : get a MWE, then clean up
+    def _contrast(self, column: str, baseline: str, group_to_compare: str) -> np.ndarray:
+        """Build a simple contrast for pairwise comparisons.
+
+        This is equivalent to
+
+        ```
+        model.cond(<column> = baseline) - model.cond(<column> = group_to_compare)
+        ```
+
+        Parameters
+        ----------
+        column: str
+            The column to contrast.
+        baseline: str
+            The baseline group.
+        group_to_compare: str
+            The group to compare to the baseline.
+
+        Returns
+        -------
+        np.ndarray
+            The contrast vector.
+        """
+        return self.cond(**{column: baseline}) - self.cond(**{column: group_to_compare})
+
+    def cond(self, **kwargs):
+        """
+        Get a contrast vector representing a specific condition.
+
+        Parameters
+        ----------
+        **kwargs
+            Column/value pairs.
+
+        Returns
+        -------
+        ndarray
+            A contrast vector that aligns to the columns of the design matrix.
+        """
+        cond_dict = kwargs
+        if not set(cond_dict.keys()).issubset(self.dds.variables):
+            raise ValueError(
+                """You specified a variable that is not part of the model. Available
+                variables: """
+                + ",".join(self.dds.variables)
             )
-            new_ref_idx = self.LFC.columns.get_loc(f"{factor}_{ref}_vs_{old_ref}")
-            self.contrast_vector[new_alternative_idx] = 1
-            self.contrast_vector[new_ref_idx] = -1
+        for var in self.dds.variables:
+            if var in cond_dict:
+                self.dds._check_category(var, cond_dict[var])
+            else:
+                cond_dict[var] = self.dds._get_default_value(var)
+        df = pd.DataFrame([kwargs])
+        return self.dds.obsm["design_matrix"].model_spec.get_model_matrix(df).iloc[0]
```

Comment:
Depends on you if you want to adopt the `.cond()` syntax for building contrasts (which was originally devised by @const-ae in [glmGamPoi](https://www.bioconductor.org/packages/release/bioc/vignettes/glmGamPoi/inst/doc/pseudobulk.html)). 

A *lot* of the code in the `_formulaic.py` module is just around finding the baseline level for each condition such that this works nicely in the case of interaction terms. 

In case you were just to support `[column, baseline, treatment]` and numpy array contrasts, you could probably come up with a way simpler solution.

Reply:
I don't have a strong opinion on this, whatever offers the most flexibility is best. 

At first I tried simplifying the code in `_formulaic.py` to keep only what I need (mainly retrieving levels for a given factor + whether it has numerical or categorical type), but I ended up keeping everything because I didn't find a straightforward simplification.

> A lot of the code in the _formulaic.py module is just around finding the baseline level for each condition such that this works nicely in the case of interaction terms.

How would you define a contrast to test interaction terms using `.cond()`? Right now I don't see how to do it without using a numerical factor.

Reply:
Let's consider a design `~ disease * timepoint`

| disease | timepoint |
|--|--|
| healthy | T0 |
| healthy | T1 |
| diseased | T0 | 
| diseased | T1 | 

which gives us the following coefficients:
`Intercept`, `diseased`, `T1`, `T1:diseased`


Then you could test:

diseased vs. healthy
```python
contrast = dds.cond(disease="diseased") - dds.cond(disease="healthy")
```

T1 vs. T0
```python
contrast = dds.cond(timepoint="T1") - dds.cond(timepoint="T0")
```

Interaction T1:diseased
```python
contrast = (
    (dds.cond(timepoint="T1", disease="diseased") - dds.cond(timepoint="T0", disease="diseased")) - 
    (dds.cond(timepoint="T1", disease="healthy") - dds.cond(timepoint="T0", disease="healthy"))
)
```

Reply:
I admit there's no suitable documentation for this in pertpy. But in principle, using this "DSL", it should be possible to specify arbitrary contrasts.